### ISSUE TITLE
DPSGD example's dataset not shuffled between epochs

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#149]
Since the dataset is turned into a list, the same batch order and batches are used for each epoch:

https://github.com/deepmind/optax/blob/30947fbc5743adc3e997c6242fa3775834862a74/examples/differentially_private_sgd.py#L131

Training loop:

https://github.com/deepmind/optax/blob/30947fbc5743adc3e997c6242fa3775834862a74/examples/differentially_private_sgd.py#L163-L168

The dataset should not be turned into a list at all, but there is a significant performance drop when the `tf.data.Dataset` is used directly (~4s/epoch with the `tf.data.Dataset`, ~0.7 seconds with the list). I was not able to improve this, so would appreciate if someone with `tf.data` expertise could take a look. Thanks!

**Comment by rosshemsley** [comment]
It is a little hidden, but the shuffle on the dataset is happening in the function that creates the dataset:

https://github.com/deepmind/optax/blob/30947fbc5743adc3e997c6242fa3775834862a74/examples/datasets.py#L59-L62
(the default value for `shuffle` is `True` here)

For using a `list()`, according to the docs for `tf.data.Dataset`, it seems this is considered idiomatic in TF2:
https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle (noting that `reshuffle_each_iteration` defaults to `True`). 

You are correct though that in the current factoring, the shuffling is deterministic across epochs (the list is only constructed once). One solution would be to construct the `list()` on each epoch.

I will look into improving this example.

**Comment by mkunesch** [comment]
Hi all! 

Thanks @n2cholas for raising this!

I agree that we should not turn the dataset into a list. I have tried removing the list and I don't observe a big performance drop on CPU and GPU. There is a small change on GPU 0.7 --> 0.9 secs per epoch but this can be solved by adding `prefetch(tf.data.AUTOTUNE)` to the datasets.

I have filed PR #230 for removing the conversion to list and adding prefetching to the datasets by default. @n2cholas would it be easy for you to check your setup with the code from the PR or could you give details about the setup you tried? No worries if not! I just wanted to make sure the performance degradation you observed is also solved by prefetching.

Thanks a lot!

**Comment by n2cholas** [comment]
Thanks @mkunesch for working on this! I still see the performance degradation with the commit you merged. On a Tesla K80 on colab, I observe 4.62 s/epoch with your latest commit, while it's about 0.41 s/epoch on an older commit. 

Here is the notebook: https://colab.research.google.com/drive/14rWb7QQSMNhNQZT_3iIBUhgQr6oH3YTo?usp=sharing. The current output is with the old commit, you can factory reset the runtime and comment out `!cd optax && git checkout 4e5ac417e99b8a6067a8d73c41b38d72e835f637 ` to see the current time.

Also, I needed to change `from optax.examples import datasets` to `import datasets` for this script to run in Colab.

**Comment by mkunesch** [comment]
Thanks a lot @n2cholas for pointing this out and for the notebook to reproduce! I'll try to figure out what the difference between the setups is ... it's quite strange that there is no slowdown for me on a similar architecture 🤔.

> Also, I needed to change from optax.examples import datasets to import datasets for this script to run in Colab.

Thanks a lot for pointing this out! This also affects the lookahead example and there is another error from jax (I think we'll have to increase the jax version requirement for the examples). I'll try to file a fix for this today or tomorrow and filed issue #240 so that we can prevent these errors in the future.

**Comment by mkunesch** [comment]
I had time to look into this now: the bottleneck was still the pre-processing on the colab GPU so caching again after preprocessing should solve the problem. 
I filed #272 and checked commit 9fe0345 with the notebook @n2cholas sent above and it runs with 0.4s/epoch again now (the same performance we had originally when we were reusing the shuffling between epochs by mistake).

**Comment by mkunesch** [comment]
I think this issue is now solved: #230 fixes the bug itself and #272 fixes the performance issue that was surfaced by the bugfix. Thanks a lot again for raising this!
